Tutorial-Learn-how-to-make-a-backend-change-that-has-an-impact-on-server-data.md

```diff
@@ -0,0 +1,677 @@
+       # Clone the model to avoid modifying the original
+       model = job_utils.clone_model(user_settings_model) 
+       # Check if the user_bio exists and is longer than 200 characters
+       if model.user_bio and len(model.user_bio) > 200:
+           # Truncate the user_bio to the first 200 characters
+           model.user_bio = model.user_bio[:200]
+           model.update_timestamps()  # Update timestamps to reflect changes
+           yield model  # Yield the modified model
+```
+
+## Section 3: Writing the Audit Job
+
+At Oppia, whenever a Beam job modifies datastore models, it is essential to write a corresponding **Audit Job**. The primary purpose of an audit job is to simulate the logic of the main job without making any actual changes to the datastore. This helps identify potential issues in the migration logic and ensures the data remains safe from unintended modifications during testing. Audit jobs are critical for maintaining confidence in the system, as they help verify the accuracy and scope of the data migration or validation process.
+
+An audit job can be used in the following ways:
+
+1. **Before the Beam Job**: To understand the scope of the data to be modified and ensure the migration logic is correct.  
+2. **After the Beam Job**: To verify that the migration was performed as expected and all affected records were updated.
+
+For instance, in the **Topic Migration Job**, the `AuditTopicMigrateJob` simulates all steps of the main `MigrateTopicJob` but does not write changes to the datastore. Similarly, we will follow this paradigm for the `TruncateUserBioJob`.
+
+The objective of our audit job, `AuditTruncateUserBioJob`, is to:
+
+1. Identify user records with bios exceeding 200 characters.  
+2. Simulate truncation logic for these records without saving the changes.  
+3. Provide a detailed report of all affected records, ensuring we are confident in the data to be modified before running the actual migration job.
+
+#### **Thought Process for the Audit Job**
+
+1. **Simulating Logic**: The audit job must simulate the exact same steps as the main Beam job to ensure consistency in logic and results.  
+2. **Read-Only Operations**: Unlike the main job, an audit job should not persist any changes to the datastore. This avoids unintended side effects during testing.  
+3. **Detailed Reporting**: The job should generate a detailed report or log indicating the records that require updates. This transparency helps validate the scope and correctness of the job.  
+4. **Reusable Patterns**: Follow established patterns and conventions for audit jobs in the Oppia codebase.
+
+> [!IMPORTANT]
+> Practice 9: Based on the above explanation and thought process. Can you write down the Audit Job for our use case.
+
+The `AuditTruncateUserBioJob` is implemented alongside the main Beam job in the `user_bio_truncation_jobs.py` file. Here’s how it can be implemented:
+
+```python
+class AuditTruncateUserBioJob(base_jobs.JobBase):
```

Comment:
One thing we generally want is for the audit and main jobs to use the same functions (rather than duplicating code). Could you give guidance on structuring the jobs to ensure that?